---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures",
  out.width = "100%"
)
```

# IOBR: Immuno-Oncology Bioinformatics Research

IOBR is an R package to perform comprehensive analysis of tumor microenvironment and signatures for immuno-oncology.

## 1 Introduction and Installation

### 1.1 Introduction

- 1.IOBR integrates 8 published methodologies decoding tumor microenvironment (TME) contexture: `CIBERSORT`, `TIMER`, `xCell`, `MCPcounter`, `ESITMATE`, `EPIC`, `IPS`, `quanTIseq`; 
- 2.IOBR collects 255 published signature gene sets, involving tumor microenvironment, tumor metabolism, m6A, exosomes, microsatellite instability, and tertiary lymphoid structure. Running the function `signature_collection_citation` to attain the source papers. The function `signature_collection` returns the detail signature genes of all given signatures.
- 3.IOBR adopts three computational methods to calculate the signature score, comprising `PCA`,`z-score`, and `ssGSEA`;
- 4.IOBR integrates multiple approaches for variable transition, visualization, batch survival analysis, feature selection, and statistical analysis.
- 5.IOBR also integrates methods for batch visualization of subgroup characteristics.

![IOBR logo](./man/figures/IOBR-Workflow.png)

![IOBR logo](./man/figures/IOBR-Package.png)

### 1.2 Installation
Before installing IOBR, please install all dependencies by executing the following command in R console:

The dependencies includs `tibble`, `survival`, `survminer`, `limma`, `limSolve`, `GSVA`, `e1071`, `preprocessCore`, `ggplot2` and `ggpubr`.

```{r}
options("repos"= c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
options(BioC_mirror="http://mirrors.tuna.tsinghua.edu.cn/bioconductor/")
if (!requireNamespace("BiocManager", quietly = TRUE)) install("BiocManager")
depens<-c('tibble', 'survival', 'survminer', 'sva', 'limma', "DESeq2","devtools",
          'limSolve', 'GSVA', 'e1071', 'preprocessCore', 'ggplot2', "biomaRt",
          'ggpubr', "devtools", "tidyHeatmap", "caret", "glmnet", "ppcor","timeROC","pracma")
for(i in 1:length(depens)){
  depen<-depens[i]
  if (!requireNamespace(depen, quietly = TRUE))
    BiocManager::install(depen)
}

if (!requireNamespace("EPIC", quietly = TRUE))
  devtools::install_github("GfellerLab/EPIC", build_vignettes=TRUE)
if (!requireNamespace("MCPcounter", quietly = TRUE))
  devtools::install_github("ebecht/MCPcounter",ref="master", subdir="Source")
if (!requireNamespace("estimate", quietly = TRUE)){
  rforge <- "http://r-forge.r-project.org"
  install.packages("estimate", repos=rforge, dependencies=TRUE)
}

```

The package is not yet on CRAN. You can install it from Github:
```{r}
if (!requireNamespace("IOBR", quietly = TRUE))
  devtools::install_github("DongqiangZeng0808/IOBR",ref="master")
```

## 1.3 Library R packages

```{r message=FALSE, warning=FALSE}
library(IOBR) 
library(EPIC)
library(estimate) 
library(MCPcounter)
Sys.setenv(LANGUAGE = "en") # The error message would be present in English.
options(stringsAsFactors = FALSE) #To avoid transiting character into factor.
```

## 1.4 Preparation of Gene-expression data derived from RNAseq.
### 1.4.1 For transcriptomic data of TCGA datasets, we strongly recomend user to use [UCSCXenaTools](https://github.com/ropensci/UCSCXenaTools) R package. Here, we download counts data of TCGA-STAD from [UCSC](https://genome.ucsc.edu/) using UCSCXenaTools R package. If you use it in published research, please cite: Wang et al., (2019). The UCSCXenaTools R package: a toolkit for accessing genomics data from UCSC Xena platform, from cancer multi-omics to single-cell RNA-seq. Journal of Open Source Software, 4(40), 1627, https://doi.org/10.21105/joss.01627
```{r}
if (!requireNamespace("UCSCXenaTools", quietly = TRUE))
    BiocManager::install("UCSCXenaTools")

library(UCSCXenaTools)
# NOTE: This process may take a few minutes which depends on the internet connection speed. Please wait for its completion.
eset_stad<-XenaGenerate(subset = XenaCohorts =="GDC TCGA Stomach Cancer (STAD)") %>% 
  XenaFilter(filterDatasets = "TCGA-STAD.htseq_counts.tsv") %>% 
  XenaQuery() %>%
  XenaDownload() %>% 
  XenaPrepare()

eset_stad[1:5,1:5]
```


### 1.4.2 Transform gene expression matrix into TPM format, and conduct subsequent annotation. 
```{r,message=FALSE,warning=FALSE}

# Remove the version numbers in Ensembl ID.
eset_stad$Ensembl_ID<-substring(eset_stad$Ensembl_ID,1,15)
eset_stad<-column_to_rownames(eset_stad,var = "Ensembl_ID")
# Revert back to original format because the data from ucsc was log2(x+1)transformed.
eset_stad<-(2^eset_stad)+1

# In this process, biomaRt R package is utilized to acquire the gene length of each Ensembl ID and calculate the TPM of each sample. If identical gene symbols exists, these genes would be ordered by the mean expression levels. The gene symbol with highest mean expression level is selected and remove others.
# NOTE: This process may take a few minutes which depends on the internet connection speed. Please wait for its completion.
eset_stad<-count2tpm(countMat = eset_stad,idType = "Ensembl", org="hsa")

eset_stad[1:5,1:5]
```

## 1.5 Preparation of Gene-expression data derived from array.

### 1.5.1 Obtain dataset of GEO gastric cancer [GSE100935](https://www.ncbi.nlm.nih.gov/pubmed/30045931) using GEOquery R package.
```{r,message=FALSE,warning=FALSE}
if (!requireNamespace("GEOquery", quietly = TRUE))
    BiocManager::install("GEOquery")
library("GEOquery")
# NOTE: This process may take a few minutes which depends on the internet connection speed. Please wait for its completion.
eset_geo<-getGEO(GEO = "GSE100935",getGPL = F,destdir = "./")

eset<-eset_geo[[1]]
eset<-exprs(eset)
eset[1:5,1:5]
```

### 1.5.2 Annotate genes in expression matrix and remove duplicate genes.
```{r,message=FALSE,warning=FALSE}

# Load the annotation file `anno_hug133plus2` in IOBR.
head(anno_hug133plus2)
# Conduct gene annotation using `anno_hug133plus2` file; If identical gene symbols exists, these genes would be ordered by the mean expression levels. The gene symbol with highest mean expression level is selected and remove others. 

eset<-anno_eset(eset = eset,
                annotation = anno_hug133plus2,
                symbol = "symbol",
                probe = "probe_id",
                method = "mean")
eset[1:5,1:5]
# Another annotation file `anno_illumina` is also provided by IOBR for convenient annotation of Illumina gene expression arrays  
head(anno_illumina)
```



## 2 TME deconvolution

### 2.1 Availabie methods to decode TME contexture

```{r}
tme_deconvolution_methods
# Return available parameter options of deconvolution methods
```

If you use this package in your work, please cite both our package and the method(s) you are using.

#### Licenses of the deconvolution methods
| method | license | citation |
|--------|---------|----------|
| [CIBERSORT](https://cibersort.stanford.edu/) | free for non-commerical use only | Newman, A. M., Liu, C. L., Green, M. R., Gentles, A. J., Feng, W., Xu, Y., … Alizadeh, A. A. (2015). Robust enumeration of cell subsets from tissue expression profiles. Nature Methods, 12(5), 453–457.  https://doi.org/10.1038/nmeth.3337 |
| [ESTIMATE](https://bioinformatics.mdanderson.org/public-software/estimate/) | free ([GPL2.0](	https://bioinformatics.mdanderson.org/estimate/)) | Vegesna R, Kim H, Torres-Garcia W, ..., Verhaak R. (2013). Inferring tumour purity and stromal and immune cell admixture from expression data. Nature Communications 4, 2612. http://doi.org/10.1038/ncomms3612 |
| [quanTIseq](http://icbi.at/software/quantiseq/doc/index.html) | free ([BSD](https://github.com/icbi-lab/immunedeconv/blob/master/LICENSE.md)) | Finotello, F., Mayer, C., Plattner, C., Laschober, G., Rieder, D., Hackl, H., ..., Sopper, S. (2019). Molecular and pharmacological modulators of the tumor immune contexture revealed by deconvolution of RNA-seq data. Genome medicine, 11(1), 34. https://doi.org/10.1186/s13073-019-0638-6 |
| [TIMER](http://cistrome.org/TIMER/) | free ([GPL 2.0](http://cistrome.org/TIMER/download.html)) | Li, B., Severson, E., Pignon, J.-C., Zhao, H., Li, T., Novak, J., … Liu, X. S. (2016). Comprehensive analyses of tumor immunity: implications for cancer immunotherapy. Genome Biology, 17(1), 174.  https://doi.org/10.1186/s13059-016-1028-7 |
| [IPS](https://github.com/icbi-lab/Immunophenogram) | free ([BSD](https://github.com/icbi-lab/Immunophenogram/blob/master/LICENSE)) | P. Charoentong et al., Pan-cancer Immunogenomic Analyses Reveal Genotype-Immunophenotype Relationships and Predictors of Response to Checkpoint Blockade. Cell Reports 18, 248-262 (2017). https://doi.org/10.1016/j.celrep.2016.12.019 |
| [MCPCounter](https://github.com/ebecht/MCPcounter) | free ([GPL 3.0](https://github.com/ebecht/MCPcounter/blob/master/Source/License)) | Becht, E., Giraldo, N. A., Lacroix, L., Buttard, B., Elarouci, N., Petitprez, F., … de Reyniès, A. (2016). Estimating the population abundance of tissue-infiltrating immune and stromal cell populations using gene expression. Genome Biology, 17(1), 218. https://doi.org/10.1186/s13059-016-1070-5 |
| [xCell](http://xcell.ucsf.edu/) | free ([GPL 3.0](https://github.com/dviraran/xCell/blob/master/DESCRIPTION)) | Aran, D., Hu, Z., & Butte, A. J. (2017). xCell: digitally portraying the tissue cellular heterogeneity landscape. Genome Biology, 18(1), 220. https://doi.org/10.1186/s13059-017-1349-1 |
| [EPIC](https://gfellerlab.shinyapps.io/EPIC_1-1/) | free for non-commercial use only ([Academic License](https://github.com/GfellerLab/EPIC/blob/master/LICENSE)) | Racle, J., de Jonge, K., Baumgaertner, P., Speiser, D. E., & Gfeller, D. (2017). Simultaneous enumeration of cancer and immune cell types from bulk tumor gene expression data. ELife, 6, e26476. https://doi.org/10.7554/eLife.26476 |

#### Licenses of the signature-esitmation method
| method | license | citation |
|--------|---------|----------|
| [GSVA](http://www.bioconductor.org/packages/release/bioc/html/GSVA.html) | free ([GPL (>= 2)](https://github.com/rcastelo/GSVA)) | Hänzelmann S, Castelo R, Guinney J (2013). “GSVA: gene set variation analysis for microarray and RNA-Seq data.” BMC Bioinformatics, 14, 7. doi: 10.1186/1471-2105-14-7, http://www.biomedcentral.com/1471-2105/14/7 |

The input data is a matrix (log2(TMP+1) transformed) containing 98 TCGA-STAD samples, with genes in rows and samples in columns. The rownames must be HGNC symbols and the colnames must be sample names.

```{r}
# Check data 
# 以上获取基因表达矩阵不成功，建议使用我们的示例数据
data("eset_stad")
eset_stad[1:5,1:5]
```

Check detail parameters of the function
```{r}
help(deconvo_tme)
```


#### 2.1.1 Method 1: Use CIBERSORT to decode the TME contexture.
```{r,fig.width= 8.5, fig.height=6}
cibersort<-deconvo_tme(eset = eset_stad, method = "cibersort",arrays = FALSE,perm = 200 )
head(cibersort)
res<-cell_bar_plot(input = cibersort[1:12,], title = "CIBERSORT Cell Fraction")
```


#### 2.1.2 Method 2: Use EPIC to decode the TME contexture.
```{r}
help(deconvo_epic)
epic<-deconvo_tme(eset = eset_stad, method = "epic",arrays = FALSE)
head(epic)

```


#### 2.1.3 Method 3: Use MCPcounter to decode the TME contexture. 
```{r}
mcp<-deconvo_tme(eset = eset_stad, method = "mcpcounter")
head(mcp)

```


#### 2.1.4 Method 4: Use xCELL to decode the TME contexture.
```{r,message=FALSE}
xcell<-deconvo_tme(eset = eset_stad, method = "xcell",arrays = FALSE)
head(xcell)

```


#### 2.1.5 Method 5: Use ESTIMATE to estimate tumor purity, and generate immune score and stromal score.
```{r,message=FALSE,warning=FALSE}
estimate<-deconvo_tme(eset = eset_stad, method = "estimate")
head(estimate)

```


#### 2.1.6 Method 6: Use TIMER to decode the TME contexture.
```{r,message=FALSE, warning=FALSE}
timer<-deconvo_tme(eset = eset_stad ,method = "timer",group_list = rep("stad",dim(eset_stad)[2]))
head(timer)

```


#### 2.1.7 Method 7: Use quanTIseq to decode the TME contexture.
```{r,fig.width= 8, fig.height=6}
quantiseq<-deconvo_tme(eset = eset_stad, tumor = TRUE, arrays = FALSE, scale_mrna = TRUE,method = "quantiseq")
head(quantiseq)
res<-cell_bar_plot(input = quantiseq[1:12,], title = "quanTIseq Cell Fraction")
```


#### 2.1.8 Method 8: Use IPS to estimate immune phenotype.
```{r,message=FALSE}
ips<-deconvo_tme(eset = eset_stad,method = "ips",plot= FALSE)
head(ips)
```


#### 2.1.9 Combine the above deconvolution results for subsequent analyses.   
```{r}
tme_combine<-cibersort %>% 
  inner_join(.,mcp,by = "ID") %>% 
  inner_join(.,xcell,by = "ID") %>%
  inner_join(.,epic,by = "ID") %>% 
  inner_join(.,estimate,by = "ID") %>% 
  inner_join(.,timer,by = "ID") %>% 
  inner_join(.,quantiseq,by = "ID") %>% 
  inner_join(.,ips,by = "ID")
dim(tme_combine)
colnames(tme_combine)
```


## 3 Signature score estimation

IOBR integrates 255 published signature gene sets, involving tumor microenvironment, tumor metabolism, m6A, exosomes, microsatellite instability, and tertiary lymphoid structure. Running the function `signature_collection_citation` to attain the source papers. The function `signature_collection` returns the detail signature genes of all given signatures.


### 3.1 Obtain the included signatures
```{r}
library(IOBR)
#TME associated signatures
names(signature_tme)[1:20]
#Metabolism related signatures
names(signature_metabolism)[1:20]
#Signatures associated with biomedical basic research: such as m6A and exosomes
names(signature_tumor)
#signature collection including all aforementioned signatures 
names(signature_collection)[1:20]

```


### 3.2 Methods for signature calculation

### 3.2.1 Calculate TME associated signatures-(through PCA method).
```{r}
sig_tme<-calculate_sig_score(pdata = NULL,
                             eset = eset_stad,
                             signature = signature_tme,
                             method = "pca",
                             mini_gene_count = 2)
sig_tme[1:5,1:10]
```

### 3.2.2 Calculate TME associated signatures-(through ssGSEA method).
```{r,message=FALSE}
sig_tme<-calculate_sig_score(pdata = NULL,
                             eset = eset_stad,
                             signature = signature_tme,
                             method = "ssgsea",
                             mini_gene_count = 5)
sig_tme[1:5,1:10]
```


### 3.2.3 Calculate metabolism related signatures.
```{r}
sig_meta<-calculate_sig_score(pdata = NULL,
                              eset = eset_stad,
                              signature = signature_metabolism,
                              method = "pca",
                              mini_gene_count = 2)
sig_meta[1:5,1:10]
```


### 3.2.4 Calculate all collected signature scores (integrating three methods: PCA, ssGSEA and z-score).
```{r,message=FALSE}
sig_res<-calculate_sig_score(pdata = NULL,
                             eset = eset_stad,
                             signature = signature_collection,
                             method = "integration",
                             mini_gene_count = 2)
sig_res[1:5,1:10]
```


### 3.2.5 The signature gene sets derived from GO, KEGG, HALLMARK and REACTOME datasets.

IOBR also enrolls the signature gene sets, containing GO, KEGG, HALLMARK and REACTOME gene sets obtained from [MsigDB](https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp#H). The ssGSEA method is recommended for these signature estimation. This process may take a while for big datasets or calculating a large number of signatures.  
```{r,message=FALSE}
sig_hallmark<-calculate_sig_score(pdata = NULL,
                                  eset = eset_stad,
                                  signature = hallmark,
                                  method = "ssgsea",
                                  mini_gene_count = 2)
sig_hallmark[1:5,1:10]
```

## 4 Find phenotype relevant signatures

### 4.1 Data of [IMvigor210 immuntherapy cohort](https://www.nature.com/articles/nature25501) is used for batch analysis of phenotype-related sigantures.
```{r}
## Load the signatures and Cell fractions calculated in previous steps by IOBR.
data("imvigor210_sig")
imvigor210_sig[1:5,1:10]

# Check therapeutic response and survival outcome in the phenotype data.
data("imvigor210_pdata")
head(imvigor210_pdata)
```

### 4.2 sig_group() function for batch visualization of response associated signatures.
```{r}
# The signature group list.
names(sig_group)

# The tumor-relevant signatures in first group.
names(sig_group)[1]
sig_group[[1]]

# The signatures associated with immunotherapy biomarkers.
names(sig_group)[3]
sig_group[[3]]

# The signatures of immunosuppression.
names(sig_group)[5]
sig_group[[5]]
```

### 4.3 Construct phenotype group and bath visualize correlation betweeen therapy response and selected signatures. 
```{r,fig.width=9,fig.height=6,warning=FALSE, message=FALSE}
pdata_group<-imvigor210_pdata[!imvigor210_pdata$BOR_binary=="NA",c("ID","BOR","BOR_binary")]
res<-iobr_cor_plot(pdata_group = pdata_group,
                   id1 = "ID",
                   feature_data = imvigor210_sig,
                   id2 = "ID",
                   target= NULL,
                   group = "BOR_binary",
                   is_target_continuous = FALSE,
                   padj_cutoff = 1,
                   index = 1,
                   category = "signature",
                   signature_group =sig_group[c(1,3,5)],
                   ProjectID = "IMvigor210",
                   palette_box = "paired1",
                   palette_corplot = "pheatmap",
                   palette_heatmap = 2,
                   feature_limit = 26,
                   character_limit = 30,
                   show_heatmap_col_name = FALSE,
                   show_col = FALSE,
                   show_plot = TRUE)
head(res,n=10)
```

### 4.4 Estemate the expression of response relevant signature genes.
```{r,fig.width=9,fig.height=6,warning=FALSE, message=FALSE}

imvigor210_eset[1:5,1:6]

res<-iobr_cor_plot(pdata_group = pdata_group,
                   id1 = "ID",
                   feature_data = imvigor210_eset,
                   id2 = "ID",
                   target= NULL,
                   group = "BOR_binary",
                   is_target_continuous = FALSE,
                   padj_cutoff = 1,
                   index = 1,
                   category = "gene",
                   signature_group =signature_collection[c(1:2,4)],
                   ProjectID = "IMvigor210",
                   palette_box = "paired1",
                   palette_corplot = "pheatmap",
                   palette_heatmap = 4,
                   feature_limit = 26,
                   character_limit = 30,
                   show_heatmap_col_name = FALSE,
                   show_col = FALSE,
                   show_plot = TRUE)
head(res,n=10)
```

## 5 Analyze signatures relevant to target signature.

### 5.1 Construct `pdata_group`as a phenotype data frame of patients.

```{r}
# The data have been saved in `imvigor210_pdata`.
head(imvigor210_pdata)
pdata_group<-as.data.frame(imvigor210_pdata[,c("ID","Pan_F_TBRs")])
pdata_group$Pan_F_TBRs<-scale(as.numeric(pdata_group$Pan_F_TBRs))
head(pdata_group)
```

### 5.2 Analyze Pan-F-TBRs relevant signatures.
`Pan-F-TBRs` is an immune-exclusion signature published in Nature in 2018: S. Mariathasan et al., TGFbeta attenuates tumour response to PD-L1 blockade by contributing to exclusion of T cells. Nature 554, 544-548 (2018). 
the Pan-F-TBRs score obtained could be used to batch analyze and visualize the interaction between this immune-exclusion signature and other interested `signatures` , as well as its significant `cell fractions`, and to further explore other biomarkers may impact corresponding biological process.



```{r,fig.width=9,fig.height=6,warning=FALSE, message=FALSE}
res<-iobr_cor_plot(pdata_group = pdata_group,
                   id1 = "ID",
                   feature_data = imvigor210_sig,
                   id2 = "ID",
                   target= "Pan_F_TBRs",
                   group = "group3",
                   is_target_continuous = TRUE,
                   padj_cutoff = 1,
                   index = 1,
                   category = "signature",
                   signature_group =sig_group[1:2],
                   ProjectID = "IMvigor210",
                   palette_box = "set2",
                   palette_corplot = "pheatmap",
                   palette_heatmap = 2,
                   feature_limit = 26,
                   character_limit = 30,
                   show_heatmap_col_name = FALSE,
                   show_col = FALSE,
                   show_plot = TRUE)
head(res,n=10)
```


### 5.3 Analyze the immune cell infiltration lanscape related to Pan-F-TBRs.
```{r,fig.width=9,fig.height=6,warning=FALSE, message=FALSE}
names(sig_group)
res<-iobr_cor_plot(pdata_group = pdata_group,
                   id1 = "ID",
                   feature_data = imvigor210_sig,
                   id2 = "ID",
                   target= "Pan_F_TBRs",
                   group = "group3",
                   is_target_continuous = TRUE,
                   padj_cutoff = 1,
                   index = 1,
                   category = "signature",
                   signature_group =sig_group[20:24],
                   ProjectID = "IMvigor210",
                   palette_box = "jco",
                   palette_corplot = "pheatmap",
                   palette_heatmap = 3,
                   feature_limit = 26,
                   character_limit = 30,
                   show_heatmap_col_name = FALSE,
                   show_col = FALSE,
                   show_plot = TRUE)
head(res,n=10)

```


## 6 Serach for mutations related to signature

### 6.1 Mutation matrix using MAF data format.
MAF data was obtained from [UCSC Xena website](https://api.gdc.cancer.gov/data/c06465a3-50e7-46f7-b2dd-7bd654ca206b)

```{r}
#添加UCSCXenaTools的下载方法
help("make_mut_matrix")

maf_file<-"TCGA.STAD.mutect.c06465a3-50e7-46f7-b2dd-7bd654ca206b.DR-10.0.somatic.maf"
# Each gene of every samples is categorized as binary variables(mutation or non-mutation) in the MAF data. 
mut_list<-make_mut_matrix(maf = maf_file,
                          isTCGA = T, 
                          category = "multi")
# If "multi" is set in above "category" parameter,four data frames will be returned, which evaluate all the mutations of every gene or estimate only SNP, indel and frameshift as follow:
names(mut_list)
mut_list$all[1:5,1:10]
```



### 6.2 sig_group() function for batch visualization of response associated signatures.
```{r,fig.width=18,fig.height=12,warning=FALSE, message=FALSE}
# Transform the mutation data into categorical variable matrix.----------
mut<-mut_list$snp
max(mut)
# If the maximum of mutation counts of a single gene of a person in is over 4, the mutation data would be standardized according to following principles.
# mut[mut>=3&mut<=5]<-3
# mut[mut>5]<-4
##########################
res<-find_mutations(mutation_matrix = mut, 
                    signature_matrix = tcga_stad_sig,
                    id_signature_matrix = "ID",
                    signature = "CD_8_T_effector",
                    min_mut_freq = 0.01,
                    plot = TRUE,
                    method = "multi",
                    save_path = paste0("CD_8_T_effector-relevant-mutations"),
                    palette = "jama",
                    show_plot = T)

```


## 7 Decode cell fraction or gene signature derived from single cell RANseq

The recent advances in single-cell analysis make it a popular alternative to determine cell markers and gene sifhatures for phenotypes. However, the significanly expensive cost and high requirement for starting tumor material still limited its widspreade utility. Therefore, the attainable bulk sequencing largr datasets is still a need in validating the clinical significance of these signatures. 

### 7.1 Method 1：Given that single-cell sequencing have a high resolution for dissecting signature genes of cells, we provide mutiple methodologies to extract cell signaturte genes from the single-cell sequencing data (`inputTPM` or `inputcounts`). Additionally, the linear` svm` algorithm of CIBERSORT or `lsei` is adopted in the bulk-seq data analysis to verify the clinical significance of thr targed cells identified by single-cell sequencing.


Data Sources: Lineage-dependent gene expression programs influence the immune landscape of colorectal cancer. Nat Genet 2020
Method: 10X 3'
Data operation process:
Cell filtering: 1. Only immune cells and stromal cells detected in the tumor are retained.
             2. Select the cell type with the cell volume greater than 500.
Gene filtering: Genes detected in more than 5% of cells.
```{r,message=FALSE,warning=FALSE}

immune_feature_limma <- GenerateRef(dat = inputTPM, pheno = pdatacounts$cluster,
                                    method = "limma", FDR = 0.05)
immune_feature_DESeq2 <- GenerateRef(dds = inputcounts,  pheno = pdatacounts$cluster, dat = inputTPM, 
                                     method = "DESeq2", FDR = 0.05)

reference <- immune_feature_DESeq2$reference_matrix
condition_number <- immune_feature_DESeq2$condition_number
top_probe <- immune_feature_DESeq2$G

```


```{r,message=FALSE,warning=FALSE}

data("eset_stad")
# reference<-reference3[,-1]
svm<-deconvo_tme(eset = eset_stad, 
                 reference = reference, 
                 method = "svm", 
                 arrays = FALSE,
                 absolute.mode = FALSE,
                 # abs.method = "sig.score",
                 perm = 1000)
head(svm)

lsei<-deconvo_tme(eset = eset_stad, 
                 reference = reference, 
                 method = "lsei", 
                 arrays = FALSE,
                 scale_reference = T,
                 perm = 1000)
head(lsei)
```


#### 7.2 Method 2：Explore the correlation between TLS score with tumor molecular characteristics, tumor microenvironment, and clinical phenotypes of patients. [B cells and tertiary lymphoid structures promote immunotherapy response](https://www.nature.com/articles/s41586-019-1922-8)


```{r,echo= FALSE, message=FALSE, warning=FALSE}
library(mydb) 
```

```{r}
# Construct the signature list as an object first. The TLs signature mentioned above is included in the 'signature_collection' in IOBR.
signature_collection$TLS_Nature

# It is recommended to calculate other included signatures simultaneously, to systematically decode the correlation between TLS score and other tumor microenvironment characteristics.

# Data of mice received checkpoint blockade is adopted in this exploration. 
(load("H:/00-IO-Db/1-RNA/3-Data/GSE63557-eset.RData"))
sig_res<-calculate_sig_score(pdata = NULL,
                             eset = eset,
                             signature = signature_collection,
                             method = "integration",
                             mini_gene_count = 2)
data(io_response)
input<-merge(io_response,sig_res,by="ID",all =FALSE)
wilcox.test(input$TLS_Nature_PCA~input$BOR_binary)

```

## 8 Other methods for batch analysis 
IOBR provide multiple batch analytical functions for statistical analysis, data operation and transformation. Batch survival analysis, for example, could batch analyze the best cutoff value for subsequent batch survival analysis, varied batch statistical tests, etc.


### 8.1 Batch survival analysis
```{r}
data("pdata_sig_tme")
input <- pdata_sig_tme %>% 
   filter(time > 0) %>% 
   filter(!is.na(status))
res <- batch_surv(pdata = input, variable = c(277:ncol(input)), time = "time",status = "status")
head(res,n=10)
```

### 8.2 Subgroup survival analyses

```{r}
##loading data and filter NA
data(subgroup_data)
input <- subgroup_data %>% 
   filter(time > 0) %>% 
   filter(!is.na(status)) %>% 
   filter(!is.na(AJCC_stage))

##for binary variable
res <- subgroup_survival(pdata = input,
                         time ="time", 
                         status = "status",
                         variable = c("ProjectID", "AJCC_stage"),
                         object ="score_binary" )
res
```

### 8.3 Batch statistical analyses

```{r,warning=FALSE,message=FALSE}
data("imvigor210_sig")
data("imvigor210_pdata")
#对于二分类变量
input<-merge(imvigor210_pdata[,c("ID","BOR_binary")],imvigor210_sig,by="ID",all = F)
input<-input[!input$BOR_binary=="NA",]
head(input)
res<-batch_wilcoxon(data = input,
                    target = "BOR_binary", 
                    feature = colnames(imvigor210_sig)[3:ncol(imvigor210_sig)])
head(res, n=10)

#对于连续性变量
head(imvigor210_sig)
res<-batch_cor(data = imvigor210_sig,
               target = "Pan_F_TBRs",
               feature = colnames(imvigor210_sig)[3:ncol(imvigor210_sig)],
               method = "spearman")
head(res, n=10)
```


### 8.4 batch_pcc的用法

例子：在矫正肿瘤纯度后跟Pan_F_TBRs相关的signature
**这一部分可以放在4 Find phenotype relevant signatures**
不足：只能算连续变量之间的偏相关系数

```{r,warning=FALSE,message=FALSE}
pdata_group <- imvigor210_pdata[, c("ID", "TumorPurity", "Pan_F_TBRs")] %>% 
  rename(target = Pan_F_TBRs) %>% mutate(target = as.numeric(target))
res <- batch_pcc(pdata_group = pdata_group, 
                 id1 = "ID", 
                 feature_data = imvigor210_sig, 
                 id2 = "ID", 
                 interferenceid = "TumorPurity",
                 target = "target",
                 method = "pearson")
head(res, n = 10)
```



### 8.5 构建模型

根据**4 Find phenotype relevant signatures**选出的marker做预后分析

```{r,warning=FALSE,message=FALSE}

pdata_group <- imvigor210_pdata[!imvigor210_pdata$BOR_binary=="NA",c("ID","BOR_binary")]
pdata_group$BOR_binary <- ifelse(pdata_group$BOR_binary == "R", 1, 0)
imvigor210_sig<-as.data.frame(imvigor210_sig)
binomial_result <- BinomialModel(x = imvigor210_sig, 
                                 y = pdata_group, 
                                 seed = "123456", 
                                 scale = TRUE,
                                 train_ratio = 0.7, 
                                 nfold = 10, 
                                 plot = T)
names(binomial_result)
lapply(binomial_result[1:3], function(z)z$AUC)

```

```{r,warning=FALSE,message=FALSE}

pdata_prog <- imvigor210_pdata[imvigor210_pdata$OS_days != "NA", c("ID","OS_days", "OS_status")]
colnames(pdata_prog) <- c("ID", "time", "status")
pdata_prog$time <- as.numeric(pdata_prog$time)
pdata_prog$status <- as.numeric(pdata_prog$status)

# PrognosticResult <- PrognosticModel(x = imvigor210_sig,
#                                     y = pdata_prog, 
#                                     scale = T, 
#                                     seed = "123456", 
#                                     train_ratio = 0.7, 
#                                     nfold = 10,
#                                     plot = T)
```




References
---------
DQ Zeng, ZL Ye, RF Shen, Y Xiong, JN Wu, WJ Qiu  WJ Liao.

IOBR: Multi-omics Immuno-Oncology Biological Research to decode tumor microenvironment and signatures.


Contact
---------
E-mail any questions to dongqiangzeng0808@gmail.com
